d2/src/main/java/com/linkedin/d2/discovery/stores/zk/ZKConnection.java

```diff
@@ -826,7 +826,21 @@ private class DefaultWatcher implements Watcher
     @Override
     public void process(WatchedEvent watchedEvent)
     {
-      ZooKeeper zk = zk();
+      ZooKeeper zk;
+      try
+      {
+        zk = zk();
+      }
+      catch (IllegalStateException e)
+      {
+        if (_shutdownAsynchronously) {
+          // On asynchronous shutdown, this can be a legitimate race.
+          LOG.debug("Watcher event received after connection shutdown (type {}, state {}.", watchedEvent.getType(),
```

Comment:
Maybe change the wording to Watched event to be consistent with the variable type?